compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEPadAndBufferize.cpp

```diff
@@ -0,0 +1,172 @@
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+static LogicalResult applyBufferizeToAllocation(RewriterBase &rewriter,
+                                                Operation *op,
+                                                Attribute memorySpace) {
+  linalg::BufferizeToAllocationOptions options;
+  options.memcpyOp =
+      linalg::BufferizeToAllocationOptions::MemcpyOp::MaterializeInDestination;
+  options.allocOp = linalg::BufferizeToAllocationOptions::AllocOp::MemrefAlloc;
+  options.bufferizeDestinationOnly = true;
+  options.emitDealloc = true;
+
+  // Bufferize ops.
+  Value buffer =
+      linalg::bufferizeToAllocation(rewriter, options, op, memorySpace);
+  if (!buffer) {
+    LLVM_DEBUG(llvm::dbgs() << "----- failed to bufferize operation -----\n");
+    return failure();
+  }
+  return success();
+}
+
+static LogicalResult applyPadAndConvertToDPS(
+    RewriterBase &rewriter, linalg::MatmulOp linalgTarget,
+    linalg::LinalgPaddingOptions &options) {
+  linalg::LinalgOp paddedOp;
+  SmallVector<Value> replacements;
+  SmallVector<tensor::PadOp> newPadOps;
+  if (failed(rewriteAsPaddedOp(rewriter, linalgTarget, options, paddedOp,
+                               replacements, newPadOps))) {
+    LLVM_DEBUG(llvm::dbgs() << "----- failed to pad op -----\n");
+    return failure();
+  }
+
+  // We need to perform our own replacement here because this API is still
+  // used in patterns that "pad and hoist", for which the replacement values
+  // need to be different.
+  // TODO: clean this up and stop "pad and hoist" behavior more globally now
+  // that we have more composable abstractions.
+  rewriter.replaceOp(linalgTarget, replacements);
+  // We rewrite each operand of the linalgOp (currently, MatmulOp) into DSP.
+  for (auto newPadOp : newPadOps) {
+    rewriter.setInsertionPointAfter(newPadOp);
+    if (failed(linalg::rewriteInDestinationPassingStyle(rewriter, newPadOp))) {
+      LLVM_DEBUG(llvm::dbgs() << "----- failed to rewrite in DPS -----\n");
+      return failure();
+    }
+  }
+
+  return success();
+}
+
+class AMDAIEPadAndBufferizePass
+    : public AMDAIEPadAndBufferizeBase<AMDAIEPadAndBufferizePass> {
+ public:
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<bufferization::BufferizationDialect, tensor::TensorDialect,
+                    linalg::LinalgDialect>();
+  }
+  void runOnOperation() override;
+};
+
+void AMDAIEPadAndBufferizePass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  IREE::HAL::ExecutableVariantOp variantOp = getOperation();
+  ModuleOp innerModule = variantOp.getInnerModule();
+  for (func::FuncOp funcOp : innerModule.getOps<func::FuncOp>()) {
+    linalg::MatmulOp matmulOp;
+    funcOp->walk<WalkOrder::PostOrder, ReverseIterator>(
+        [&](TilingInterface op) {
+          // Find the next matmul op if it does not have loops.
+          if (op.getLoopIteratorTypes().empty() || !isa<linalg::MatmulOp>(op))
```

Comment:
Maybe make this an `LinalgOp`

Reply:
Hmmm, this was never actually done and is the reason for issues here, https://github.com/nod-ai/iree-amd-aie/pull/206
@Abhishek-Varma did you make a stab at it and ran into issues, or just didnt have a testr case yet so skipped. I am asking becuase I will try to do it tomorrow so just pre-fetching what you might know.

Reply:
You can disregard was pretty trivial to add it, will push a commit with this change by EOD.